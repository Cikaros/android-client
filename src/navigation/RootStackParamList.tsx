export type RootStackParamList = {
  Home: undefined;
  Onboarding: undefined;
  ChangeServer: undefined;
  LeftDrawer: undefined;
  About: undefined,
  Advance: undefined,
  Peers: undefined,
};
