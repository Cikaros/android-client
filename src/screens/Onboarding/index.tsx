import React, {useEffect} from 'react';
import {useDispatch} from 'react-redux';
import {Image, StyleSheet, Text, View} from 'react-native';
import {RootStackParamList} from '@navigation/RootStackParamList';
import {StackScreenProps} from '@react-navigation/stack';
import Button from '../../components/Button/index';
import {actions as appStatusActions} from '../../store/app-status';
import SplashScreen from 'react-native-splash-screen';
import {actions as appStatusPersistActions} from '../../store/app-status-persist';

type Props = StackScreenProps<RootStackParamList, 'Onboarding'>;

const Onboarding: React.FC<Props> = ({navigation}) => {
  const dispatch = useDispatch();
  useEffect(() => SplashScreen.hide(), []);

  const onContinue = (showChangeServer: boolean) => {
    dispatch(appStatusActions.setShowLeftDrawer(true));
    dispatch(appStatusActions.setTitleMainNavigator(''));
    dispatch(appStatusPersistActions.setOnboardingShowed(true));
    dispatch(appStatusActions.setOnboardingToChangeServer(showChangeServer));
    navigation.navigate('Home');
  };

  return (
    <View
      style={{
        flex: 1,
        justifyContent: 'center',
        backgroundColor: '#F2F2F2',
      }}>
      <View style={styles.Wrapper}>
        <View style={{alignItems: 'center'}}>
          <Image
            style={styles.HeroImage}
            source={require('@assets/images/logo-onboarding.png')}
          />
        </View>
        <Text style={styles.Title}>
          <Text>
            By default you will connect to {'\n'} NetBird's cloud servers. You
            can {'\n'} access the{' '}
          </Text>
          <Text
            testID="123"
            style={styles.TitleLink}
            onPress={() => onContinue(true)}>
            change server
          </Text>
          <Text> menu to use {'\n'} another server.</Text>
        </Text>
        <Button onPress={() => onContinue(false)} text="Continue" />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  HeroImage: {
    width: 118,
    marginBottom: 88,
    resizeMode: 'contain',
  },
  Wrapper: {
    paddingHorizontal: 41,
    paddingVertical: 41,
  },
  container: {
    paddingHorizontal: 0,
    paddingVertical: 41,
    alignItems: 'center',
  },
  Title: {
    textAlign: 'center',
    marginBottom: 38,
    paddingHorizontal: 0,
    paddingVertical: 2,
    lineHeight: 22,
    color: '#000000',
    fontSize: 16,
    fontWeight: '500',
    fontFamily: 'Roboto',
  },
  TitleLink: {
    color: '#F68330',
    fontSize: 16,
    fontWeight: '500',
    fontFamily: 'Roboto',
    textDecorationLine: 'underline',
  },
});

export default Onboarding;
