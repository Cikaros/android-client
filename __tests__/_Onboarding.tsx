import React from 'react';
import {shallow, mount, render} from 'enzyme';
import Onboarding from '../src/screens/Onboarding';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
// import {actions as appStatusActions} from '../src/store/app-status'; 
// import {actions as appStatusPersistActions} from '../src/store/app-status-persist';
import {RootStackParamList} from '@navigation/RootStackParamList';
import {StackNavigationProp} from '@react-navigation/stack';
import {RouteProp} from '@react-navigation/native';

const navigation = {
  navigate: jest.fn(),
  goBack: jest.fn(),
  dispatch: jest.fn(),
  reset: jest.fn(),
  isFocused: jest.fn(),
  canGoBack: jest.fn(),
} as unknown as StackNavigationProp<RootStackParamList, 'Onboarding'>;

const route = {
  key: 'test-key',
  name: 'Onboarding',
} as RouteProp<RootStackParamList, 'Onboarding'>;

const hideMock = jest.fn();
jest.mock('react-native-splash-screen', () => ({hide: jest.fn()}));

const mockStore = configureStore();
const store = mockStore({});

describe('Onboarding component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <Onboarding navigation={navigation} route={route} />
      </Provider>,
    );
    expect(wrapper.exists()).toBe(true);
  });

  it('should hide the splash screen on mount', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <Onboarding navigation={navigation} route={route} />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });

//   it('should dispatch the correct actions when "Continue" is pressed', () => {
//     const setShowLeftDrawerMock = jest.fn();
//     const setTitleMainNavigatorMock = jest.fn();
//     const setOnboardingShowedMock = jest.fn();
//     const setOnboardingToChangeServerMock = jest.fn();

//     jest
//       .spyOn(appStatusActions, 'setShowLeftDrawer')
//       .mockImplementation(setShowLeftDrawerMock);
//     jest
//       .spyOn(appStatusActions, 'setTitleMainNavigator')
//       .mockImplementation(setTitleMainNavigatorMock);
//     jest
//       .spyOn(appStatusPersistActions, 'setOnboardingShowed')
//       .mockImplementation(setOnboardingShowedMock);
//     jest
//       .spyOn(appStatusActions, 'setOnboardingToChangeServer')
//       .mockImplementation(setOnboardingToChangeServerMock);

//     const wrapper = shallow(
//       <Provider store={store}>
//         <Onboarding navigation={navigation} route={route} />
//       </Provider>,
//     );

  
//       // Simulate button press
//       const button = wrapper.find('Onboarding').dive().find(`Button[text="Continue"]`);
//       button.simulate('press');

//     // Assertions on dispatched actions and other functionality
//     expect(setOnboardingShowedMock).toHaveBeenCalledWith();
//     expect(setOnboardingToChangeServerMock).toHaveBeenCalledWith(false);
//     expect(navigation.navigate).toHaveBeenCalledWith('Home');
//     expect(setShowLeftDrawerMock).toHaveBeenCalledWith(true);
//     expect(setTitleMainNavigatorMock).toHaveBeenCalledWith('');
//     // expect(dispatchMock).toHaveBeenCalledTimes(4);
//   });

  //   it('should dispatch the correct actions when "Continue" is pressed', () => {
  //     const setShowLeftDrawerMock = jest.fn();
  //     const setTitleMainNavigatorMock = jest.fn();
  //     const setOnboardingShowedMock = jest.fn();
  //     const setOnboardingToChangeServerMock = jest.fn();

  //     // Mock the useDispatch hook
  //     const dispatchMock = jest.fn();
  //     jest.mock('react-redux', () => ({
  //       useDispatch: () => jest.fn(),
  //     }));

  //     const wrapper = shallow(
  //       <Provider store={store}>
  //         <Onboarding navigation={navigation} route={route} />
  //       </Provider>,
  //     );
  //     const button = wrapper.find(`Button[text="Continue"]`);
  //     // expect(button).toExist();
  //     button.simulate('press');

  //     const instance = wrapper.instance() as Onboarding;
  //     instance.onPress();

  //     // const button = wrapper
  //     // .findWhere(node => node.prop('testID') === '123')
  //     //   .first();

  //     //   button.prop('onClick');

  //     jest
  //       .spyOn(appStatusActions, 'setShowLeftDrawer')
  //       .mockImplementation(setShowLeftDrawerMock);

  //     jest
  //       .spyOn(appStatusActions, 'setTitleMainNavigator')
  //       .mockImplementation(setTitleMainNavigatorMock);
  //     jest
  //       .spyOn(appStatusPersistActions, 'setOnboardingShowed')
  //       .mockImplementation(setOnboardingShowedMock);
  //     jest
  //       .spyOn(appStatusActions, 'setOnboardingToChangeServer')
  //       .mockImplementation(setOnboardingToChangeServerMock);
  //     // expect(setShowLeftDrawerMock).toHaveBeenCalledWith(true);
  //     // expect(setTitleMainNavigatorMock).toHaveBeenCalledWith('');
  //     expect(setOnboardingShowedMock).toHaveBeenCalledWith(true);
  //     expect(setOnboardingToChangeServerMock).toHaveBeenCalledWith(false);
  //     expect(mockNavigation.navigate).toHaveBeenCalledWith('Home');
  //     expect(dispatchMock).toHaveBeenCalledTimes(4);
  //   });

  //   it('should dispatch the correct actions when "change server" link is pressed', () => {
  //     const setShowLeftDrawerMock = jest.fn();
  //     const setTitleMainNavigatorMock = jest.fn();
  //     const setOnboardingShowedMock = jest.fn();
  //     const setOnboardingToChangeServerMock = jest.fn();
  //     const dispatchMock = jest.fn();

  //     // Mock the useDispatch hook to return the mocked dispatch function
  //     jest.mock('react-redux', () => ({
  //         useDispatch: () => jest.fn(),
  //         }));

  //         const wrapper = shallow(
  //             <Provider store={store}>
  //               <Onboarding navigation={mockNavigation} />
  //             </Provider>
  //           );
  //         const titleLink = wrapper.find('Text').findWhere((node) => node.props().id === '123');

  //     // Set the mocked functions as implementations for the dispatch actions
  //     jest.spyOn(appStatusActions, 'setShowLeftDrawer').mockImplementation(setShowLeftDrawerMock);
  //     jest.spyOn(appStatusActions, 'setTitleMainNavigator').mockImplementation(setTitleMainNavigatorMock);
  //     jest.spyOn(appStatusPersistActions, 'setOnboardingShowed').mockImplementation(setOnboardingShowedMock);
  //     jest.spyOn(appStatusActions, 'setOnboardingToChangeServer').mockImplementation(setOnboardingToChangeServerMock);

  //     // Simulate link press
  //     titleLink.props().onPress();

  //     expect(setShowLeftDrawerMock).toHaveBeenCalledWith(true);
  //     expect(setTitleMainNavigatorMock).toHaveBeenCalledWith('');
  //     expect(setOnboardingShowedMock).toHaveBeenCalledWith(true);
  //     expect(setOnboardingToChangeServerMock).toHaveBeenCalledWith(true);
  //     expect(mockNavigation.navigate).not.toHaveBeenCalled(); // Should not navigate to 'Home'
  //     expect(dispatchMock).toHaveBeenCalledTimes(4);
  //   });

  // You can add more test cases for other scenarios if needed
});
