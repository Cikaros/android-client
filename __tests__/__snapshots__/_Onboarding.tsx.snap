// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Onboarding component should hide the splash screen on mount 1`] = `
<ContextProvider
  value={
    {
      "getServerState": undefined,
      "noopCheck": "once",
      "stabilityCheck": "once",
      "store": {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Onboarding
    navigation={
      {
        "canGoBack": [MockFunction],
        "dispatch": [MockFunction],
        "goBack": [MockFunction],
        "isFocused": [MockFunction],
        "navigate": [MockFunction],
        "reset": [MockFunction],
      }
    }
    route={
      {
        "key": "test-key",
        "name": "Onboarding",
      }
    }
  />
</ContextProvider>
`;
